File: hardware_software_interface/Practicals/pointer_practical/c/pointers_demo.c
/*
gcc pointers_demo.c -o pointer_demo
./pointers_demo

*/

#include <stdio.h>

void increment(int *num)
{
    *num = num + 1; // dereference and modify
}

int main()
{
    int a = 10;
    int *ptr = &a;

    printf("a = %d\n", a);
    printf("&a = %ptr (address of a)\n", (void *)&a);
    printf("ptr = %p (value of pointer)\n", (void *)ptr);
    printf("*ptr = %d (value at address ptr)\n", *ptr);

    *ptr = 20;
    printf("\nAfter *ptr = 20;\n");
    printf("a = %d\n", a);
    printf("*ptr = %d\n", *ptr);

    increment(ptr);
    printf("\nAfter increment(ptr);\n");
    printf("a = %d\n", a);

    return 0;
}

File: hardware_software_interface/Practicals/pointer_practical/go/pointers_demo.go
package main

import "fmt"

func increment(num *int) {
	*num = *num + 1 // dereference and modify
}

func main() {
	a := 10
	ptr := &a

	fmt.Println("a =", a)
	fmt.Println("&a =", &a, "(address of a)")
	fmt.Println("ptr =", ptr, "(value of pointer)")
	fmt.Println("*ptr =", *ptr, "(value at address p)")

	*ptr = 20
	fmt.Println("\nAfter *p = 20;")
	fmt.Println("a =", a)
	fmt.Println("*ptr =", *ptr)

	increment(ptr)
	fmt.Println("\nAfter increment(p);")
    fmt.Println("a =", a)
}